### ISSUE TITLE
Very long and redundant file-path

### ISSUE THREAD

**Issue by dostuffthatmatters** [Lith/strapi-provider-upload-google-cloud-storage#38]
**Describe the bug**
When uploading a file with the standard configuration (you provide) the resulting file location in the configured bucket is very long: Every image is located in its own subdirectory and the file name includes the filename + the filehash, whereas the filehash would be just fine.

**Example (See the image below):**
An image `DSC2737.jpeg` with the hash `DSC2737_2cab11f170` gets stored under `<bucket-name>/DSC2737_2cab11f170/_DSC2737_DSC2737_2cab11f170.jpeg`. Why not just `<bucket-name>/DSC2737_2cab11f170.jpeg`?

**Screenshots**
<img width="1534" alt="bug_report_1" src="https://user-images.githubusercontent.com/29046316/82560323-800a7480-9b71-11ea-8999-d6612f988737.png">

**Comment by dgmike** [comment]
I think the problem is focuses in the part where the developer can do a upload referenced media to strapi as described in documentation: https://strapi.io/documentation/3.0.0-beta.x/plugins/upload.html#upload-files-related-to-an-entry

With this approach the upload file name goes well formatted. But in normal way, using admin panel or simple upload form, the upload file name really goes so trick: it repeats the file hash in path, including size (thumbnail, large...) as parent path and the first part is "undefined_".

**Comment by Lith** [comment]
Not sure next release will fix that with `basePath`.
I reopened it, but this "problem" provide from Strapi Upload plugin their self.

**Comment by dgmike** [comment]
Looking at how the official providers work, I realized that the files are clustered in the root folder, without adding any type of change.

AWS S3:

```js
          const path = file.path ? `${file.path}/` : '';
          S3.upload(
            {
              Key: `${path}${file.hash}${file.ext}`,
              Body: Buffer.from(file.buffer, 'binary'),
```

Cloudinary:

```js
        const upload_stream = cloudinary.uploader.upload_stream(
            { resource_type: 'auto', public_id: file.hash, ...customConfig },
            (err, image) => {
```

Local file:

```js
          // write file in public/assets folder
          fs.writeFile(
            path.join(strapi.config.paths.static, `/uploads/${file.hash}${file.ext}`),
            file.buffer,
```

Rackspace:

```js
        const writeStream = client.upload({
          ...options,
          remote: file.hash,
          contentType: file.mime,
        });
```

All of then uses `file.hash` and in some cases `file.ext` attributes to store the files. Maybe it could be a good approach to use.

**Comment by dgmike** [comment]
@Lith @dostuffthatmatters What's your opinion about that? :point_up_2: :

**Comment by dostuffthatmatters** [comment]
@dgmike Thank you for looking this up! Looks like Cloudinary and Rackspace get the extension implicitly from the uploaded file itself.

I will try to modify the `generateUploadFileName` method (https://github.com/Lith/strapi-provider-upload-google-cloud-storage/blob/master/lib/provider.js#L118) tonight.

<br/>

**Impact on current users**
The impact of file path changes for old users is definitely something we'd have to look out for!

We could maybe use an environment variable `GOOGLE_CLOUD_STORAGE_SHORT_PATHS` (or something similar) and only if that one exists and is set to true then the mentioned method uses the new shorter path.

I could write a script that turns the old urls into the new urls (database and bucket), I will use one for MongoDB anyway. However a simple "Replace media" in the strapi dashboard would also do it.

**Comment by dgmike** [comment]
@dostuffthatmatters I preffer to use configurations over environment variables because is the default way to configure providers. And, to configure it on strapi could be using environment variable.

I don't really know about the possible problems with script and interfere on strapi dashboard for a provider. Maybe two `PR` could be done with each solutions for avail it...

PS: now we have a contribution guide. Is good to take a look before start. :wink:

**Comment by dostuffthatmatters** [comment]
@dgmike Yes, you are right: Config vars would be more elegant!

The scripting thing is not something I would recommend to a regular user 😅 But I already did that in a project and the media collection in the database as well as the actual files are the only things you'd have to modify.

I guess with a proper way to set this new method via the configuration with the default config being the old way there is no need for two `PR` right?

**Comment by shabbir21** [comment]
Hi team,
Any update/timeline for the above?

**Comment by ScottEAdams** [comment]
I have cloned the library locally and use:
```
const generateUploadFileName = (basePath, file) => {
  const path = file.path ? `${file.path}/` : '';
  return `${basePath}${path}${file.hash}${file.ext}`;
};
```
And it is working exactly as I would expect it to. i.e. if uploading via admin it goes straight into basePath and if uploading via multipart I can set the sub path of base like aws plugin.

**Comment by Lith** [comment]
I've miss to check for this feature 😱

Sorry guys, and thanks @ScottEAdams for your code improvements.
Send me a PR, I will check it asap.

Strapi Admin panel do not use path configuration, so... this great feature is only available for custom client side application.

I will add it on next Strapi release.

Thanks

**Comment by Lith** [comment]
Fixed in release 3.1.6
